compiler/rustc_ty_utils/src/layout.rs

```diff
@@ -80,7 +80,36 @@ fn layout_of<'tcx>(
 }
 
 fn error<'tcx>(cx: &LayoutCx<'tcx>, err: LayoutError<'tcx>) -> &'tcx LayoutError<'tcx> {
-    cx.tcx.arena.alloc(err)
+    cx.tcx().arena.alloc(err)
+}
+
+fn map_error<'tcx>(
+    cx: &LayoutCx<'tcx>,
+    ty: Ty<'tcx>,
+    err: LayoutCalculatorError,
+) -> &'tcx LayoutError<'tcx> {
+    let err = match err {
+        LayoutCalculatorError::SizeOverflow => {
+            // This is sometimes not a compile error in `check` builds.
+            LayoutError::SizeOverflow(ty)
+        }
+        LayoutCalculatorError::UnexpectedUnsized => {
+            // This is sometimes not a compile error if there are trivially false where
+            // clauses, but it is always a compiler error in the empty environment.
+            if cx.param_env.caller_bounds().is_empty() {
+                cx.tcx().dcx().delayed_bug(format!(
+                    "encountered unexpected unsized field in layout of {ty:?}"
+                ));
+            }
```

Comment:
We could perhaps attempt to prove that the type is sized here.

Reply:
Also would be useful to point to an example.

Reply:
> We could perhaps attempt to prove that the type is sized here.

The final type is not always guaranteed to be `Sized` (in the type system way), for example given
```rust
#![feature(trivial_bounds)]
struct Foo
where
    [u8]: Sized,
{
    a: [u8],
    b: [u16],
}
```

the trait bound `Foo: Sized` will not hold, but attempting to compute the layout of `Foo` will error with `UnexpectedUnsized`.

Reply:
Right, but shouldn't we be able to prove that the type is sized within the param-env we're computing the layout? Or otherwise, why would we care that the param-env has no clauses?

Reply:
Like, I'd want to double check that the param-env in fact does contain `[u8]: Sized` rather than just checking that the param-env is non-empty.

Reply:
Right, makes sense to check for the field itself. I'll try to add this check.

Reply:
Ok, I've replaced the check for an empty param-env with a check that the layout-unsized field is type-system-sized now.

Also, looking at the code again, I found that an old hack could now be replaced, so I did that in another commit.